packages/yarnpkg-pnpify/sources/hoist.ts

```diff
@@ -0,0 +1,283 @@
+ * @param packages package infos
+ *
+ * @returns package ids with max weights among the packages with the same name
+ */
+const getHeaviestPackages = (weights: ReadonlyWeightMap, packages: ReadonlyArray<HoisterPackageInfo>): Set<HoisterPackageId> => {
+  const heaviestPackages = new Map<HoisterPackageName, {weight: HoisterWeight, pkgId: HoisterPackageId}>();
+  for (const [pkgId, weight] of weights) {
+    const pkgName = packages[pkgId].name;
+    let heaviestPkg = heaviestPackages.get(pkgName);
+    if (!heaviestPkg) {
+      heaviestPkg = {weight, pkgId};
+      heaviestPackages.set(pkgName, heaviestPkg);
+    } else if (weight > heaviestPkg.weight) {
+      heaviestPkg.weight = weight;
+      heaviestPkg.pkgId = pkgId;
+    }
+  }
+
+  const heavyPackageIds = new Set<HoisterPackageId>();
+  for (const {pkgId} of heaviestPackages.values())
+    heavyPackageIds.add(pkgId);
+
+  return heavyPackageIds;
+};
+
+/**
+ * Find the packages that can be hoisted to the subtree root `rootId`.
+ *
+ * @param tree package tree
+ * @param rootId package id that should be regarded as subtree root
+ * @param packages package infos
+ * @param nohoist nohoist package ids
+ */
+const computeHoistCandidates = (tree: HoisterPackageTree, rootId: HoisterPackageId, packages: ReadonlyArray<HoisterPackageInfo>, nohoist: ReadonlySet<HoisterPackageId>): Set<HoisterPackageId> => {
+  // Get current package dependency package names
+  const rootDepNames = new Map<HoisterPackageName, HoisterPackageId>();
+  for (const depId of tree[rootId].deps)
+    rootDepNames.set(packages[depId].name, depId);
+
+  // Weigh all the packages in the subtree
+  const packageWeights = weighPackages(tree, rootId, packages, nohoist);
+
+  const hoistCandidateWeights: WeightMap = new Map();
+  const hoistPeerDepCandidates = new Set<HoisterPackageId>();
+  const seenPackageNames = new Set<HoisterPackageName>();
+  const seenPackageIds = new Set<HoisterPackageId>();
+
+  const findHoistCandidates = (nodeId: HoisterPackageId) => {
+    seenPackageIds.add(nodeId);
+    const name = packages[nodeId].name;
+    // Package names that exist only in a single instance in the tree path are hoist candidates
+    if (!seenPackageNames.has(name)) {
+      seenPackageNames.add(name);
+      const rootDepId = rootDepNames.get(name);
+      // If the hoisting candidate has the same name as existing root subtree dependency,
+      // we can only hoist it if its id is also the same
+      // . → A → B@X → C → B@Y, - we can hoist only B@X here
+      if (nodeId !== rootId && (!rootDepId || rootDepId === nodeId)) {
+        const pkg = tree[nodeId];
+        if (pkg.peerDeps.size > 0) {
+          hoistPeerDepCandidates.add(nodeId);
+        } else {
+          hoistCandidateWeights.set(nodeId, packageWeights.get(nodeId)!);
+        }
+      }
+
+      for (const depId of tree[nodeId].deps)
+        if (!seenPackageIds.has(depId))
+          findHoistCandidates(depId);
```

Comment:
I think there's a problem here - you need to proceed in three passes (kinda like we do in virtual package resolution): first you add all the children to `seenPackageIds`, then you iterate inside them, then you remove them.

In the current setup, if `root` depends on `A` and `B`, and if `A` depends on `B`, then `B` will be marked as a hoist candidate.

Reply:
There is no problem here, discussed on discord. The `depId` is a name + version, and the checks for keeping require and peer dependency promises are done in the code above.